libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/api/MapboxManeuverApi.kt

```diff
@@ -0,0 +1,119 @@
+import com.mapbox.navigation.ui.base.api.maneuver.StepDistanceRemainingCallback
+import com.mapbox.navigation.ui.base.api.maneuver.UpcomingManeuversCallback
+import com.mapbox.navigation.ui.base.model.maneuver.Maneuver
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState.CurrentManeuver
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState.DistanceRemainingToFinishStep
+import com.mapbox.navigation.ui.maneuver.ManeuverAction
+import com.mapbox.navigation.ui.maneuver.ManeuverProcessor
+import com.mapbox.navigation.ui.maneuver.ManeuverResult
+import com.mapbox.navigation.utils.internal.JobControl
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of [ManeuverApi] allowing access to generate representation of
+ * maneuvers based on [BannerInstructions]
+ * @property distanceFormatter DistanceFormatter
+ */
+class MapboxManeuverApi internal constructor(
+    private val distanceFormatter: DistanceFormatter,
+    private val processor: ManeuverProcessor
+) : ManeuverApi {
+
+    /**
+     * @param formatter contains various instances for use in formatting distance related data
+     * for display in the UI
+     *
+     * @return a [MapboxManeuverApi]
+     */
+    constructor(formatter: DistanceFormatter) : this(formatter, ManeuverProcessor)
+
+    private val mainJobController: JobControl by lazy { ThreadController.getMainScopeAndRootJob() }
+    private var currentManeuverJob: Job? = null
+    private var upcomingManeuverJob: Job? = null
+
+    /**
+     * Given [BannerInstructions] the method goes through the list of [BannerComponents] and returns
+     * a [Maneuver] wrapped inside [ManeuverCallback]
+     * @param bannerInstruction BannerInstructions object representing [BannerInstructions]
+     * @param callback ManeuverCallback contains [ManeuverState.CurrentManeuver]
+     */
+    override fun retrieveManeuver(
+        bannerInstruction: BannerInstructions,
+        callback: ManeuverCallback
+    ) {
+        currentManeuverJob = mainJobController.scope.launch {
+            val action = ManeuverAction.ParseCurrentManeuver(bannerInstruction)
+            val result = processor.process(action) as ManeuverResult.CurrentManeuver
+            callback.onManeuver(CurrentManeuver(result.currentManeuver))
+        }
+    }
+
+    /**
+     * Given [RouteStepProgress] the method returns the distance remaining to finish the [LegStep]
+     * @param routeStepProgress RouteStepProgress progress object specific to the current step the user is on
+     * @param callback StepDistanceRemainingCallback contains [ManeuverState.DistanceRemainingToFinishStep]
+     */
+    override fun retrieveStepDistanceRemaining(
+        routeStepProgress: RouteStepProgress,
+        callback: StepDistanceRemainingCallback
+    ) {
+        mainJobController.scope.launch {
```

Comment:
Same as above. (cancel is missing for this one entirely)

Reply:
I didn't think we would need one for this because it is not doing any network operations. If the idea is to have for the sake of consistency I can add one.

Reply:
Not only consistency but also leaks. If the `callback` has UI references, and we take a long time to process, the activity might be dead by the time we return which would leak resources.